src/Microsoft.Framework.BufferedHtmlContent.Sources/BufferedHtmlContent.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.AspNet.Html.Abstractions;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.Framework.Internal
+{
+    /// <summary>
+    /// Enumerable object collection which knows how to write itself.
+    /// </summary>
+    internal class BufferedHtmlContent : IHtmlContent, IEnumerable<object>
+    {
+        private const int MaxCharToStringLength = 1024;
+
+        // This is not List<IHtmlContent> because that would lead to wrapping all strings to IHtmlContent
+        // which is not space performant.
+        // internal for testing.
+        internal List<object> Entries { get; } = new List<object>();
+
+        /// <summary>
+        /// Appends the string to the collection.
+        /// </summary>
+        /// <param name="value">The <c>string</c> to be appended.</param>
+        public BufferedHtmlContent Append(string value)
+        {
+            Entries.Add(value);
+            return this;
+        }
+
+        /// <summary>
+        /// Appends a character array to the collection.
+        /// </summary>
+        /// <param name="value">The character array to be appended.</param>
+        /// <param name="index">The index from which the character array must be read.</param>
+        /// <param name="count">The count till which the character array must be read.</param>
+        /// <remarks>
+        /// Splits the character array into strings of 1KB length and appends them.
+        /// </remarks>
+        public BufferedHtmlContent Append([NotNull] char[] value, int index, int count)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            if (count < 0 || value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = count;
+                if (MaxCharToStringLength < currentCount)
+                {
+                    currentCount = MaxCharToStringLength;
+                }
+
+                Append(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Appends a <see cref="IHtmlContent"/> to the collection.
+        /// </summary>
+        /// <param name="htmlContent">The <see cref="IHtmlContent"/> to be appended.</param>
+        public BufferedHtmlContent Append(IHtmlContent htmlContent)
```

Comment:
How much does performance improve if you special case appending another `BufferedHtmlContent` instance as `BufferEntryCollection` did before -- appending the passed instance's entries rather than sticking the `BufferedHtmlCotnent` itself into the collection? That would avoid some recursion in `WriteTo()` and allow the old instance to be garbage collected earlier.